### PR TITLE
kubeadm: enable kubelet client certificate rotation on primary CP nodes

### LINKED ISSUE #1753: kubeadm should leverage kubelet automatic client cert rotation on nodes created with `kubeadm init`
## Is this a BUG REPORT or FEATURE REQUEST?

FEATURE REQUEST

## Versions

**kubeadm version** `kubeadm version: &version.Info{Major:"1", Minor:"15", GitVersion:"v1.15.3", GitCommit:"2d3c76f9091b6bec110a5e63777c332469e0cba2", GitTreeState:"clean", BuildDate:"2019-08-26T15:59:52Z", GoVersion:"go1.12.5", Compiler:"gc", Platform:"darwin/amd64"}`

## What happened?
On master nodes, `/etc/kubernetes/kubelet.conf` gets created with "hardcoded" `client-certificate/client-key` instead of pointing to `/var/lib/kubelet/pki/kubelet-client-current.pem` as done on minions node.

## What you expected to happen?
I expected `/etc/kubernetes/kubelet.conf` to point to `/var/lib/kubelet/pki/kubelet-client-current.pem` to leverage automatic kubelet client certificate rotation that is configured by `kubeadm`

## How to reproduce it (as minimally and precisely as possible)?
`kubeadm init && cat /etc/kubernetes/kubelet.conf`

## Anything else we need to know?
I already know this a chicken-and-egg problem but I think it would be really nice if the first master, after initialising the control plane, could make use of `/var/lib/kubelet/pki/kubelet-client-current.pem` to further streamline the certificates rotation process and avoid having to use `kubeadm init kubeconfig kubelet` just on the master nodes to renew kubelet's client certificate.

### PR DESCRIPTION
**Pull request by neolit123**
**What this PR does / why we need it**:

- Don't always print to stdout that the kubelet is starting.
instead delegate this to the callers of TryStartKubelet.
- Add a new root kubeadm init phase called kubelet-finalize
- Add a sub-phase to "kubelet-finalize" called "cert-rotation"
- "cert-rotation" performs the following actions:
  - tries to guess if kubelet client cert rotation is enabled
  - update the kubelet.conf to use the rotatable cert/key

This is an alternative approach to the one we tried here:
https://github.com/kubernetes/kubernetes/pull/83339
(but it was reverted https://github.com/kubernetes/kubernetes/pull/84012)

See https://github.com/kubernetes/kubeadm/issues/1753#issuecomment-525068877 for more details.

**Which issue(s) this PR fixes**:

Fixes kubernetes/kubeadm#1753

**Special notes for your reviewer**:
NONE
**Does this PR introduce a user-facing change?**:

```release-note
ACTION REQUIRED: kubeadm: add a new "kubelet-finalize" phase as part of the "init" workflow and an experimental sub-phase to enable automatic kubelet client certificate rotation on primary control-plane nodes.

Prior to 1.17 and for existing nodes created by "kubeadm init" where kubelet client certificate rotation is desired, you must modify "/etc/kubernetes/kubelet.conf" to point to the PEM symlink for rotation:
"client-certificate: /var/lib/kubelet/pki/kubelet-client-current.pem" and "client-key: /var/lib/kubelet/pki/kubelet-client-current.pem", replacing the embedded client certificate and key.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

/assign @fabriziopandini 
/priority important-soon
/kind bug
@kubernetes/sig-cluster-lifecycle-pr-reviews

### CODE REVIEW

File: cmd/kubeadm/app/cmd/init.go
Comment (neolit123) [inline cmd/kubeadm/app/cmd/init.go]: these changes above are auto-cleanup of whitespace.

File: cmd/kubeadm/app/cmd/phases/init/kubelet.go
Comment (neolit123) [inline cmd/kubeadm/app/cmd/phases/init/kubelet.go]: for init and join moving the "starting kubelet" message on the /cmd side.
[1]

File: cmd/kubeadm/app/cmd/phases/join/kubelet.go
Comment (neolit123) [inline cmd/kubeadm/app/cmd/phases/join/kubelet.go]: [1]

File: cmd/kubeadm/app/phases/kubelet/kubelet.go
Comment (neolit123) [inline cmd/kubeadm/app/phases/kubelet/kubelet.go]: [1]

File: cmd/kubeadm/app/cmd/phases/init/kubeletready.go
Comment (neolit123) [inline cmd/kubeadm/app/cmd/phases/init/kubeletready.go]: the new phase names are debatable. opinions?

Comment (fabriziopandini) [reply cmd/kubeadm/app/cmd/phases/init/kubeletready.go]: no strong opinions, but since we already have `kubelet-[verb]` for `kubelet-start`, and almost all nouns for the existing sub-phases, probably we should follow the same pattern also here. e.g. 
```
kubelet-complete
    certificate-rotation (or cert-rotation)
```

File: cmd/kubeadm/app/cmd/phases/init/kubeletready.go
Comment (fabriziopandini) [inline cmd/kubeadm/app/cmd/phases/init/kubeletready.go]: Forgive my bad memory, but what is the other sub-phase we are planning to add here?

File: cmd/kubeadm/app/cmd/phases/init/kubeletready.go
Comment (fabriziopandini) [inline cmd/kubeadm/app/cmd/phases/init/kubeletready.go]: Only for double-check. 
Is it correct to have both `ClientKey` and  `ClientCertificate` pointing to the same file?

**Review by rosti — APPROVED** [review]
Thanks @neolit123 !
Only minor nits on the current implementation.
I do feel, however, that we should go with the bootstrap config (your previous PR) in the long run. But let's use this one for now.
/lgtm
/hold
for discussion. Feel free to remove the hold when ready.

**Review by fabriziopandini — COMMENTED** [review]
@neolit123 Thanks for working on this issue!
I like much better this solution because it is explicit what is doing and why.
I have left few small questions, but overall lgtm.

Let's give some time to for other feedbacks as well

File: cmd/kubeadm/app/cmd/phases/init/kubeletready.go
Comment (neolit123) [reply cmd/kubeadm/app/cmd/phases/init/kubeletready.go]: I think this differnt on the windows service. Then again there is no init on windows yet.

File: cmd/kubeadm/app/cmd/phases/init/kubeletready.go
Comment (neolit123) [reply cmd/kubeadm/app/cmd/phases/init/kubeletready.go]: Whitelist pod labels

File: cmd/kubeadm/app/cmd/phases/init/kubeletready.go
Comment (neolit123) [reply cmd/kubeadm/app/cmd/phases/init/kubeletready.go]: If its a pem that has both inside, its ok

File: cmd/kubeadm/app/cmd/phases/init/kubeletready.go
Comment (neolit123) [reply cmd/kubeadm/app/cmd/phases/init/kubeletready.go]: Sgtm. Comments by others?

**Comment by fabriziopandini** [comment]
Let's try to get a final agreement on phase name today at the office hours

File: cmd/kubeadm/app/cmd/phases/init/kubeletready.go
Comment (rosti) [reply cmd/kubeadm/app/cmd/phases/init/kubeletready.go]: Nope, it's the same everywhere. `ServiceRestart` exists on Windows, it's called and it should work as expected (despite all it does is a stop, then start).

File: cmd/kubeadm/app/cmd/phases/init/kubeletready.go
Comment (fabriziopandini) [reply cmd/kubeadm/app/cmd/phases/init/kubeletready.go]: kubelet-finalize

File: cmd/kubeadm/app/cmd/phases/init/kubeletready.go
Comment (neolit123) [reply cmd/kubeadm/app/cmd/phases/init/kubeletready.go]: ack, just saw there is ServiceRestart for both OSes.

File: cmd/kubeadm/app/cmd/phases/init/kubeletready.go
Comment (ereslibre) [reply cmd/kubeadm/app/cmd/phases/init/kubeletready.go]: @rosti, will it? If it's really `nil`, indexing on `nil` will fail on runtime, unless it's being defaulted to an empty map.

Comment (neolit123) [reply cmd/kubeadm/app/cmd/phases/init/kubeletready.go]: i think it works like @rosti says: https://play.golang.org/p/ZQ80fXE7azG

Comment (ereslibre) [reply cmd/kubeadm/app/cmd/phases/init/kubeletready.go]: Yes, when having the map type works as expected 👍

**Comment by neolit123** [comment]
@fabriziopandini @rosti @ereslibre i have updated this PR>

**Review by ereslibre — APPROVED** [review]
Thank you @neolit123.

/lgtm
/retest
/hold
^ for other reviewers, feel free to unhold.

A question: do we plan to enable certificate rotation upon `kubeadm upgrade apply/node` for existing clusters?

**Comment by neolit123** [comment]
> A question: do we plan to enable certificate rotation upon kubeadm upgrade apply/node for existing clusters?

that's a good question.
i'm +1, but it is a bit tricky.

`kubeadm upgrade apply` can be called on any CP node and not only on the node where `kubeadm init` was called. the problem only exists on that node. so this means that all subcommands on `kubeadm upgrade` need to know about this fix and ideally the fix has to be temporary in the 1.17 binary; removed in the 1.18 binary.

the fix would also need the --config which should contain a KubeletConfiguration to determine if the user is using a custom kubelet PKI directory, to be able to establish if certificate rotation is enabled.

and finally this would also require to move the logic that was added in the new phase in the kubeadm "backend", so that it can also be consumed during upgrade commands.

**Comment by neolit123** [comment]
@fabriziopandini @rosti 
ping for review before code-freeze.

**Comment by fabriziopandini** [comment]
I'm +1 to get this merged
/approve
/lgtm

**Comment by ereslibre** [comment]
/hold cancel
^ based on discussion.

**Comment by rosti** [comment]
Reapply the hold and check if upgrade fixes this one
/hold

**Comment by neolit123** [comment]
confirmed that the kubelet.conf still contains embedded / outdated certificates after upgrade.
slack discussion:
https://kubernetes.slack.com/archives/C13J86Z63/p1573495392079000

File: cmd/kubeadm/app/cmd/phases/init/kubeletfinalize.go
Comment (rosti) [inline cmd/kubeadm/app/cmd/phases/init/kubeletfinalize.go]: We may as well prefix this with `experimental-` as well. If we decide to keep it we can rename it and add an alias to the old name. As long as all of the sub-phases of `kubelet-finalize` are marked as experimental, then that can be `experimental-` prefixed too.
It's best to decide as soon as possible if we want to merge this or not. I am in for merging as long as we can change things in the future (remove the phase and opt in for another solution).
@neolit123 @fabriziopandini WDYT?

Comment (neolit123) [reply cmd/kubeadm/app/cmd/phases/init/kubeletfinalize.go]: updating with experimental and also using EXPERIMENTAL in the description because it's better than alpha.

Comment (neolit123) [reply cmd/kubeadm/app/cmd/phases/init/kubeletfinalize.go]: the experimental- pattern seems fine in terms of giving us the option to remove things.
we have no other way...

**Review by rosti — APPROVED** [review]
Thanks @neolit123 !
/lgtm

We probably need to mention in the release note, that the position and name of the phase might change, hence it's marked as alpha.

**Comment by neolit123** [comment]
> We probably need to mention in the release note, that the position and name of the phase might change, hence it's marked as alpha.

the "experimental" part already gives us this freedom.

**Comment by fabriziopandini** [comment]
/approve
/lgtm
/hold cancel
@neolit123 after this merge, we update kinder to run this phase in our E2E

**Pull request merged**
